### PR TITLE
extractors: restructuring classes

### PR DESCRIPTION
**Pull request by awillecke**
Restructuring of classes to remove duplicate code due to the introduction of SqlExtractor

* renamed old BaseExtractor to OmnetExtractor
* New BaseExtractor implements basic reading of SQLite files and category conversion
* SqlExtractor inherits from new BaseExtractor
* BaseExtractor can now include the filename in the returned DataFrame

### CODE REVIEW

**Comment by hagau** [comment]
I've rewritten the `convert_columns_to_category` (and renamed it to the more fitting `convert_columns_dtype`) to only convert explicitly given column to either categorical or numerical data. 
- The types of columns that are not mentioned are chosen by pandas automatically
- numerical columns can be converted to a specified type via a given dictionary
- `examples/AoI.yaml` has been updates as a more complex example
- unfortunately, upon reading the data back in in `PlottingReaderFeather`, the conversion has to be done again to save on memory usage. I've added the two parameters there and the example in `AoI.yaml` uses anchors and aliases to alleviate the need to repeat the same unwieldy data structure.
- I've only tested the `PatternMatchingBulkExtractor` via the AoI example, but I'm fairly confident everything else should work too. I'll test the other extractors while cleaning up the recipes :slightly_smiling_face:

File: exporters.py [hunk 1]
```diff
@@ -141,8 +141,7 @@ def prepare_concatenated(self, data_list, job_list):
             job = dask.delayed(self.save_to_disk)(map(operator.itemgetter(0), data_list), self.output_filename, self.format)
         else:
             concat_result = dask.delayed(pd.concat)(map(operator.itemgetter(0), data_list), ignore_index=True)
-            convert_columns_result = dask.delayed(RawExtractor.convert_columns_to_category)(concat_result)
-            job = dask.delayed(self.save_to_disk)(convert_columns_result, self.output_filename, self.format)
+            job = dask.delayed(self.save_to_disk)(concatenate, self.output_filename, self.format)
```
Comment (awillecke) [inline exporters.py]: I think this should be `concat_result` instead of `concatenate` right?

**Comment by awillecke** [comment]
Thank you for implementing explicit type conversion! It already works great. Currently, I'm testing a few recipes and will report back.

File: exporters.py (hunk 1 above)
Comment (hagau) [reply exporters.py]: Yes, thank you! Sorry, autocomplete tricked me there and I only tested the `prepare_separated` pathway :blush: 
I've pushed a fix now.

**Comment by awillecke** [comment]
I've tested the `SqlExtractor` and `RawExtractor` and both work as expected :+1: 

Also, I've added the possibility to have regex in every part of files added as `input_files` to allow something like this when data is structured in subdirectories:
```
evaluation: !Evaluation
  extractors:
  - dataset: !SqlExtractor
      input_files: !!python/list
        - "/tmp/ns3/experiment_[a-z].*/scenario_[a-z].*/run_[0-9].*.db"
```

**Comment by hagau** [comment]
@awillecke Thanks, that's way more useful than the rather limited functionality that existed before.
I've generalized the used regular expression to allow for paths like `/tmp/ns3/experiment._[a-z].*/scenario_[a-z].*/run_[0-9].*.db` (note the first dot). This will still match paths with e.g. escaped parentheses as part of the regex, but that will not result in any problems in the selection.
I've also converted the docstrings to numpydoc format. I will add a comment in the documentation that it is the format for docstrings used throughout the code.

**Comment by hagau** [comment]
> @awillecke Thanks, that's way more useful than the rather limited functionality that existed before. I've generalized the used regular expression to allow for paths like `/tmp/ns3/experiment._[a-z].*/scenario_[a-z].*/run_[0-9].*.db` (note the first dot). This will still match paths with e.g. escaped parentheses as part of the regex, but that will not result in any problems in the selection. I've also converted the docstrings to numpydoc format. I will add a comment in the documentation that it is the format for docstrings used throughout the code.

I have to add that for things like escaped parentheses in the path, one has to use a plain or single-quoted scalar in YAML. Unfortunately, I've used the double-quoted style in most examples. I'm going to add a comment in the documentation and a couple of examples in the example recipes to make that more clear.

**Comment by awillecke** [comment]
Thank you for review. I've not considered that case.

**Comment by awillecke** [comment]
IMO it is intuitive to set all columns in the `categorical_columns` and `numerical_columns` parameters that are expected to be the output of the extractor.  Also, sometimes a user would like to convert tags to other data types. 

Then it makes sense to call `convert_columns_dtype` after the whole data frame is built (so after applying the tags). Hence, I moved the call for type conversion to `read_query_from_file` which now extracts data via `read_sql_from_file` and applies tags in the case of `OmnetExtractor` first, before setting the respective types.

**Comment by hagau** [comment]
> IMO it is intuitive to set all columns in the `categorical_columns` and `numerical_columns` parameters that are expected to be the output of the extractor. Also, sometimes a user would like to convert tags to other data types.
> 
> Then it makes sense to call `convert_columns_dtype` after the whole data frame is built (so after applying the tags). Hence, I moved the call for type conversion to `read_query_from_file` which now extracts data via `read_sql_from_file` and applies tags in the case of `OmnetExtractor` first, before setting the respective types.

Thanks, that makes more sense.

**Comment by hagau** [comment]
I've extended the column dtype conversion to apply a more sensible ordering on the elements. Previously, the ordering was automatically derived and lexical for strings (e.g. 'pdr_3' came after 'pdr_29'), now a relatively simple heuristic is used to determine the type contained in a string, it's split into tuples and those tuples are then used to produce a proper order.

I've also added back the type conversion when exporting to allow for conversion after a transform. One could argue that the transform are a better place for that, since the point of the conversion is mostly to conserve memory, but that would be a separate PR, I think. The call to the conversion function is also much cheaper now, so multiple calls to it for different columns are not a problem.

**Comment by hagau** [comment]
I will test and adapt more of the example recipes later on, but I think this is ready to be merged :+1:

**Comment by awillecke** [comment]
I also think it's fine to merge now :+1:

**Pull request merged**